pkg/model/gcemodel/autoscalinggroup.go

```diff
@@ -226,3 +257,191 @@ func (b *AutoscalingGroupModelBuilder) Build(c *fi.ModelBuilderContext) error {
+		for k, vr := range it.Metadata {
+			if k == "startup-script" {
+				b, err := vr.AsBytes()
+				if err != nil {
+					return nil, err
+				}
+				bootstrapData = b
+				continue
+			}
+
+			v, err := vr.AsString()
+			if err != nil {
+				return nil, err
+			}
+
+			additionalMetadata[k] = v
+		}
+
+		// TODO: GCP provider requires this, but actually only requires it when we aren't using a custom image
+		version := cluster.Spec.KubernetesVersion
+
+		var machineTemplate *unstructured.Unstructured
+		{
+			u := &unstructured.Unstructured{}
+
+			u.SetAPIVersion("infrastructure.cluster.x-k8s.io/v1alpha3")
+			u.SetKind("GCPMachineTemplate")
+			u.SetName(ig.Name)
+			u.SetNamespace(ig.Namespace)
+
+			// TODO: We should support instances on GCP without a public IP
+			// (I think we need to set up a NAT gateway)
+			publicIP := true
+
+			template := map[string]interface{}{
+				"spec": map[string]interface{}{
+					"instanceType":          it.MachineType,
+					"zone":                  zone,
+					"image":                 image,
+					"rootDeviceSize":        it.BootDiskSizeGB,
+					"serviceAccount":        serviceAccount,
+					"publicIP":              &publicIP,
+					"additionalNetworkTags": it.Tags,
+					"additionalMetadata":    additionalMetadata,
+				},
+			}
+
+			spec := map[string]interface{}{
+				"template": template,
+			}
+
+			u.Object["spec"] = spec
+			machineTemplate = u
+		}
+
+		// We use a hash to ensure that we pick up new versions and avoid mutating in-place
+		// TODO: Use hash from kustomize?
+		hash := fmt.Sprintf("%x", sha256.Sum256(bootstrapData))[0:6]
+
+		dataSecretName := ig.Name + "-bootstrap-" + hash
+
+		var bootstrapSecret *unstructured.Unstructured
+		{
+			u := &unstructured.Unstructured{}
+
+			u.SetAPIVersion("v1")
+			u.SetKind("Secret")
+			u.SetName(dataSecretName)
+			u.SetNamespace(ig.Namespace)
+
+			data := map[string]interface{}{
+				"value": base64.StdEncoding.EncodeToString([]byte(bootstrapData)),
+			}
+
+			u.Object["data"] = data
+			u.Object["type"] = "Opaque"
+
+			bootstrapSecret = u
+		}
+
+		var machineDeployment *unstructured.Unstructured
+		{
+			u := &unstructured.Unstructured{}
+
+			u.SetAPIVersion("cluster.x-k8s.io/v1alpha3")
+			u.SetKind("MachineDeployment")
+			u.SetName(ig.Name)
+			u.SetNamespace(ig.Namespace)
+
+			labels := map[string]string{
+				"kops.k8s.io/instancegroup": ig.Name,
+			}
+
+			/*
+				// For reasons not entirely clear, we resolve the cluster from the machine using this label
+				// TODO: Use owner refs if available?
+				labels["cluster.x-k8s.io/cluster-name"] = cluster.Name
+			*/
+
+			template := map[string]interface{}{
+				"metadata": map[string]interface{}{
+					"labels": labels,
+				},
+				"spec": map[string]interface{}{
+					"clusterName": cluster.Name,
+					"bootstrap": map[string]interface{}{
+						"dataSecretName": dataSecretName,
+					},
+					"infrastructureRef": map[string]interface{}{
```

Comment:
I assume you're using these maps to avoid the dependency of Cluster API repo, right?

Reply:
Right - it's actually not so much a circular dependency as it is that cluster-api has its own dependencies on controller-runtime & client-go etc and I was hitting conflicts.